src/dump.c

```diff
@@ -2023,15 +2194,117 @@ static void jl_insert_methods(jl_array_t *list)
     size_t i, l = jl_array_len(list);
     for (i = 0; i < l; i += 2) {
         jl_method_t *meth = (jl_method_t*)jl_array_ptr_ref(list, i);
+        assert(jl_is_method(meth));
         assert(!meth->is_for_opaque_closure);
         jl_tupletype_t *simpletype = (jl_tupletype_t*)jl_array_ptr_ref(list, i + 1);
-        assert(jl_is_method(meth));
         jl_methtable_t *mt = jl_method_get_table(meth);
         assert((jl_value_t*)mt != jl_nothing);
         jl_method_table_insert(mt, meth, simpletype);
     }
 }
 
+void remove_code_instance_from_validation(jl_code_instance_t *codeinst)
+{
+    ptrhash_remove(&new_code_instance_validate, codeinst);
+}
+
+static void jl_insert_method_instances(jl_array_t *list)
+{
+    size_t i, l = jl_array_len(list);
+    // Validate the MethodInstances
+    jl_array_t *valids = jl_alloc_array_1d(jl_array_uint8_type, l);
+    memset(jl_array_data(valids), 1, l);
+    size_t world = jl_atomic_load_acquire(&jl_world_counter);
+    for (i = 0; i < l; i++) {
+        jl_method_instance_t *mi = (jl_method_instance_t*)jl_array_ptr_ref(list, i);
+        assert(jl_is_method_instance(mi));
+        if (jl_is_method(mi->def.method)) {
+            // Is this still the method we'd be calling?
+            jl_methtable_t *mt = jl_method_table_for(mi->specTypes);
+            jl_value_t *mworld = jl_methtable_lookup(mt, mi->specTypes, world);
```

Comment:
This is probably not the right method query to use; it will only find exact matches. Plus some specializations might be from `invoke`s.

Reply:
Should we just call `jl_typemap_assoc_by_type` directly with `subtype=1`? Or is there another high-level caller you'd recommend? See also https://github.com/JuliaLang/julia/blob/1b47efb08753a808c4869bb4fe0561e526cde612/src/dump.c#L2621-L2628 which seems a bit duplicative.

Reply:
Good catch on the `invoke` though. Looks like we don't even handle that correctly now: with

```julia
module PkgA

f(::Int) = 1
f(::Any) = 2

g() = invoke(f, Tuple{Any}, 0)

g()

end # module
```
and
```
module PkgB

using PkgA

PkgA.f(::Real) = 3

end # module
```
we still invalidate `g()` when loading `PkgB`

```julia
julia> using PkgA

julia> PkgA.g()
2

julia> m = which(PkgA.g, ())
g() in PkgA at /tmp/pkgv/PkgA/src/PkgA.jl:6

julia> mi = m.specializations[1]
MethodInstance for PkgA.g()

julia> mi.cache.max_world
0xffffffffffffffff

julia> using PkgB

julia> mi.cache.max_world
0x0000000000007a61
```
This PR will make the same mistake, but I don't think it's worse than the status quo.

Reply:
I filed this as #44320